## ALL IMPORTING DONE HERE


In [24]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

### LOADING DATA

In [25]:
df=pd.read_csv('online_retail.csv')
df.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom


In [26]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  str    
 1   StockCode    541909 non-null  str    
 2   Description  540455 non-null  str    
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  str    
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  str    
dtypes: float64(2), int64(1), str(5)
memory usage: 33.1 MB


### we have loaded and checked  , we have got an info about the dataset, while it seems there may null values in Customer_id

### -- before handling null we may eliminate duplicate rows

## CHECKING DUPLIACTE ROWS

In [27]:
df.duplicated().sum()

np.int64(5268)

### -- it means there are 5268 duplicate rows

### -- then we should delete the duplicated rows by only keeping the first row only 

In [28]:
df.drop_duplicates(keep='first', inplace=True)

#### -- here we have used drop_dupliactes to remove the duplicate row , keep='first' -- dentoes to keep one row from the duplicate ,i,e store one row form duplicated,, inplace=true denote do it in inplace

### -now we again recheck using df.duplicated.sum()

In [29]:
df.duplicated().sum()

np.int64(0)

### now we have ZERO duplicated rows

## NULL VALUES 

In [30]:
df.isnull().sum()

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135037
Country             0
dtype: int64

### there are many null values here

### --- CUSTOMER ID HAVE 13K NULL VALUES AND THAT CLOUMN HAVE NO CONNECTING WITH TARGET ,SO DROPING IT IS THE BEST APPROACH

In [31]:
df.drop(columns=['CustomerID'], inplace=True)

### --even DESCRIPTON HAVE MORE NULL VALUE AND NO CONNECTION WITH TARGET , SOO DROPING IS THE BEST APPROACH

In [32]:
df.drop(columns=['Description'], inplace=True)

### WE HAVE DROPED THE NULL VALUE , ONCE WE MIGTH CHECK IS THERE ANY NULL VALUE AGAIN

In [33]:
df.isnull().sum()

InvoiceNo      0
StockCode      0
Quantity       0
InvoiceDate    0
UnitPrice      0
Country        0
dtype: int64

### THEREFORE NO NULL VALUES , WE HAVE HANDLED NULL VALUES 

## OUTLIER DETECTION